File: src/node/app/tote.js
'use strict'

module.exports = class Tote {
    constructor(betPools) {
        this.poolsStakesPerRunner = {};
        this.poolsTotalStake = {};
        this.betPools = {};

        betPools.forEach(function (betPool) {
            this.poolsStakesPerRunner[betPool.type] = {};
            this.poolsTotalStake[betPool.type] = 0;
            this.betPools[betPool.type] = betPool;
        }, this);
    }

    addBet(type, runner, stake) {
        if (typeof this.betPools[type] === 'undefined') {
            throw new Error("Bet pool does not exist");
        }

        if (!this.poolsStakesPerRunner[type].hasOwnProperty(runner)) {
            this.poolsStakesPerRunner[type][runner] = 0;
        }

        this.poolsStakesPerRunner[type][runner] = this.poolsStakesPerRunner[type][runner] + stake;
        this.poolsTotalStake[type] = this.poolsTotalStake[type] + stake;
    }

    totalStakeForType(type) {
        if (typeof this.betPools[type] === 'undefined') {
            throw new Error("Bet pool does not exist");
        }

        return this.poolsTotalStake[type];
    }

    totalStakeForRunner(type, runner) {
        if (typeof this.betPools[type] === 'undefined') {
            throw new Error("Bet pool does not exist");
        }

        return this.poolsStakesPerRunner[type][runner] || 0;
    }

    getResult(type, runner) {
        if (typeof this.betPools[type] === 'undefined') {
            throw new Error("Bet pool does not exist");
        }

        var betPool = this.betPools[type];
        var stakeForRunner = this.totalStakeForRunner(type, runner);

        if (stakeForRunner == 0) {
            return 0;
        }

        return Number((this.totalStakeForType(type) * betPool.stakeFactor / stakeForRunner).toFixed(2));
    }

    getResults(first, second, third) {
        var $this = this;

        var result = {
            first: first,
            second: second,
            third: third
        };

        var results = [];

        Object.keys(this.betPools).forEach(function(key,index) {
            var betPool = $this.betPools[key];
            betPool.results = [];

            betPool.results = betPool
                .resultsFor
                .map(function (position) {
                    return $this.getResult(betPool.type, position(result));
                });
            
            results.push(betPool);
        });

        return results;
    }
}

File: src/dotnet/ToteChallenge/Domain/Tote.cs
﻿using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;

namespace ToteChallenge.Domain
{
    public class Tote
    {
        private readonly IDictionary<string, BetPool> _betPools;

        private readonly Dictionary<string, ConcurrentDictionary<string, decimal>> _poolsStakesPerRunner = new Dictionary<string, ConcurrentDictionary<string, decimal>>();
        private readonly ConcurrentDictionary<string, decimal> _poolsTotalStake = new ConcurrentDictionary<string, decimal>();

        public bool Open { get; private set; }

        public Tote(IEnumerable<BetPool> betPools)
        {
            _betPools = betPools
                .ToDictionary(betPool => betPool.Type);

            Open = true;

            InitialisePools();
        }

        private void InitialisePools()
        {
            foreach (var betPool in _betPools)
            {
                _poolsStakesPerRunner.Add(betPool.Value.Type.ToLower(), new ConcurrentDictionary<string, decimal>());
                _poolsTotalStake.AddOrUpdate(betPool.Value.Type, 0, (oldKey, oldValue) => 0);
            }
        }

        public void AddBet(string type, string runner, decimal stake)
        {
            if (!_betPools.ContainsKey(type))
            {
                throw new ArgumentException("Bet pool does not exist.");
            }

            var runnerStakes = _poolsStakesPerRunner[type];
            runnerStakes.AddOrUpdate(runner, stake, (key, current) => current + stake);

            _poolsTotalStake.AddOrUpdate(type, stake, (key, current) => Math.Round(current + stake));
        }

        public decimal TotalStakeForType(string type)
        {
            if (!_betPools.ContainsKey(type))
            {
                throw new ArgumentException("Bet pool does not exist.");
            }

            return _poolsTotalStake[type];
        }

        public decimal TotalStakeForRunner(string type, string runner)
        {
            if (!_betPools.ContainsKey(type))
            {
                throw new ArgumentException("Bet pool does not exist.");
            }

            return _poolsStakesPerRunner[type].ContainsKey(runner)
                ? _poolsStakesPerRunner[type][runner]
                : 0;
        }

        public decimal GetResult(string type, string runner)
        {
            if (!_betPools.ContainsKey(type))
            {
                throw new ArgumentException("Bet pool does not exist.");
            }

            var betPool = _betPools[type];
            var totalStakeForRunner = TotalStakeForRunner(type, runner);

            if (totalStakeForRunner == 0)
            {
                return 0;
            }

            return Math.Round(TotalStakeForType(type) * betPool.StakeFactor / totalStakeForRunner, 2);
        }

        public void CloseTote()
        {
            Open = false;
        }
    }
}